family-dashboard/api/src/main/java/org/phenotips/familydashboard/internal/OmimToHpoMapper.java

```diff
@@ -0,0 +1,248 @@
+{
+    private List<String> omimTerms;
+    private List<String> hpoTerms;
+    private Map<String, Map<String, JSONObject>> omimToHpoMapping;
+    private JSONObject origObj;
+    private JSONObject currentMatch;
+    private final int matchThreshold;
+
+    protected Vocabulary omimService;
+    protected Vocabulary hpoService;
+
+    public OmimToHpoMapper(JSONObject data, int matchThreshold, Vocabulary omimService, Vocabulary hpoService)
+    {
+        this.origObj = data;
+        this.matchThreshold = matchThreshold;
+        this.omimService = omimService;
+        this.hpoService = hpoService;
+        this.setDisorders(data);
+        this.setFeatures(data);
+        this.setOmimToHpoMapping();
+    }
+
+    /**
+     * Updates the family member's phenotype list with those that are relevant to the omim disorder(s) the
+     * member is diagnosed with.
+     *
+     */
+    public void updateFamilyMemberVocabularies()
+    {
+        updateFamilyMemberDisorders();
+        updateFamilyMemberPhenotypes();
+    }
+
+    private void updateFamilyMemberPhenotypes()
+    {
+        this.origObj.putOpt("features", this.getRelevantPhenotypes());
+    }
+
+    private void updateFamilyMemberDisorders()
+    {
+        this.origObj.putOpt("disorders", this.getOmimDisorders());
+    }
+
+    private JSONObject getOmimTermResults(String term)
+    {
+        VocabularyTerm omimTerm = this.omimService.getTerm(term);
+        if (omimTerm != null) {
+            return omimTerm.toJSON();
+        }
+        return null;
+    }
+
+    /**
+     * This method checks:
+     *  1\ Whether the actual symptom for an OMIM disorder matches with a patient's phenotype set
+     *  2\ Whether a parent term to the actual symptom matches with a patient's phenotype set
+     *  3\ Whether a parent term to each item in a patient's phenotype set matches with the actual symptom
+     * If any of the above are true, the term from the patient's phenotype set which produced a match gets saved.
+     *
+     * @param symptom the actual symptom that is associated with an OMIM disorder
+     * @return true if a match between any of the above 3 cases are found, and false otherwise
+     */
+    private boolean hasRelevantPhenotype(String symptom)
+    {
+        if (this.hpoTerms.contains(symptom)) {
+            setCurrentMatch(this.hpoService.getTerm(symptom));
+            return true;
+
+        } else if (foundPatientPhenotypesToSymptomParentsMatch(symptom)) {
+            return true;
+
+        } else if (foundSymptomToPatientPhenotypeParentsMatch(symptom)) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Checks for a match between the parent terms of the actual_symptom with any item in the patient's phenotype set.
+     *
+     * @param symptom the actual symptom that is associated with an OMIM disorder
+     * @return
+     */
+    private boolean foundPatientPhenotypesToSymptomParentsMatch(String symptom)
```

Comment:
Should only direct parents be considered? Another idea would be to check if it is among all ancestors, and if yes, check if the distance to it is at most 3.